### DISCUSSION TITLE
M5Stack core 2

### DISCUSSION THREAD

**Discussion by prschguy1** [PlummersSoftwareLLC/NightDriverStrip#420] [Q&A]
Wanted to see if anyone has tried building Spectrumstack using M5Stack core 2 package? I have been working with it a bit over the weekend. Seems it will build without any errors listed. Seems like most of it works very well, and is quite impressive. It is mostly identical to my M5stick, just bigger/better. Very good job on this. Display, VU meter, sound react, touch screen all working. When I visit the ip address, statistics, and effects all look good to go. Unfortunately, it for whatever reason fails to output pin 0. Not sure why that is happening. Pin0 is defined in several different locations that all seem to be correct. One thing I noticed is that I am not seeing specific globals for this project, so I'm not sure where it is getting it's settings from. I tried to assign it a set of globals to it, however studio code was very unhappy with that. Was hoping someone might have some ideas on what to try.

Thanks, Craig

**Comment by davepl** [comment]
Normally for an M5 product LED_PIN0 will wind up defined as 26 or 33, depending if it’s an M5 Stick C, Stick C Plus, or Stack.  Shouldn’t be 0 unless you redefined it?

Check which pin in on your external port, and define it to be that, and it should work!

**Comment by prschguy** [comment]
Hello Dave, thank you for the reply. I guess i should have more clear in my description. In my setup pin0  =32   which is the original setup. I also have 33 available in this board. Changing it to that didnt help either. Oddly, if i make a new environment ledstrip_stack, then i do get data from pin 32 enough to run a strip. Thats why i was curious where it gets its settings from.

Thanks, craig

**Comment by davepl** [comment]
Here’s the code of interest in globals.h

    #if SPECTRUM_WROVER_KIT
        #define LED_PIN0        5
    #elif ELECROW
        #define LED_PIN0        19
    #else
        #define LED_PIN0        26
    #endif


Without SPECTRUM_WROVER_KIT or ELECROW defined, you should get 26 here.  What does LED_PIN0 actually wind up set to for you?

**Comment by prschguy1** [comment]
Hello Dave, as always, you were right on point with your answer. I had thought those settings had resided here 
 271       #if M5STICKC || M5STICKCPLUS || M5STACKCORE2
272        #define LED_PIN0 32

had changed port number here to get m5stick plus working.
unfortunately, I hadn't scrolled down far enough to find 

 838      #if SPECTRUM_WROVER_KIT
 839      #define LED_PIN0        5
 840      #elif ELECROW
 841      #define LED_PIN0        19
 842      #else
 843      #define LED_PIN0        32
 844      #endif

so, again you are right. Thank you so much for answering that for me. saved me a lot of time. everything working as intended now. As they  say the devil is in the details. Now that I know this unit works in strip form, just might pursue that further. While this chip set is rather pin short, seems really stable. would rather use esp32 -s3 devkit c, but that's not so stable right now. So will be patient on that right now.

So will say thanks again,   Craig

**Comment by robertlipe** [comment]
Pro tip when chasing a define: just before it's used, throw away a build
with #defune THE THING 0x4321 or something crazy.

The error you'll get will tell you where the canonical deine die THETHING
is.

I have a CL in process for a diy-stule "raw"  sep32-s3 board with 8mb psram
and 16MB flash that's quite stable. Psram was a little tricky, but other
boards are using it so the parts are mostly there.

**Comment by davepl** [comment]
Glad it’s working for you, and pleased to have been of help!

- Dave

**Comment by prschguy1** [comment]
![20230828_181611_resized](https://github.com/PlummersSoftwareLLC/NightDriverStrip/assets/69419979/d8e04f50-ff0a-408f-ab53-a84c140ff8bb)

Thank you Robert for the tip. I enjoy reading your posts here. I have learned a lot from you so far. You have a tendency to think outside of the box. Something I think I need to do more of. I am more of a hardware guy, and programming is new to me. While I used to know my way around this project pretty well, there has been a lot of refactoring, and I am finding I'll just have to re-read a lot of the files here to understand what does what now. Probably time to start just trying stuff to see what works, and what doesn't.

Thanks again, Craig

**Comment by robertlipe** [comment]
OOooh, Aaaah. That's nice!

You have extremely square pixels for a WS2812 grid. Is there some kind of
overlay/mask involved that keeps all the light from bleeding together or is
this a swanky fixture? That fancy connector at the bottom and the sheath
looks pretty nice compared to the mess I always end up with. ;-)

Thank you for the kind words. I tend to write (admittedly, sometimes, too
much) based on my own exploration of this code and merging that with
decades of software experience and a hardware background of my own. I'm
comfortable reading code and helping others. I've been doing it in some
form since the 1200bps modem days. There's lots to learn from each other in
projects like this and being able to work with people that aren't just
assigned to a neighboring cell block, err, cubicle lets you connect with
all kind of fun and smart people in OpenSourceVille. When a body failure
forced involuntary retirement, I picked the part of my career I enjoyed and
tinkering with electronics/systems software/group projects is how I fill my
days (and nights) these days.

I just had a bunch of 16x16 WS2812's delivered here last week, but two
panels were defective. (Ali was pretty quick with the refund, but now I
have to put more on the proverbial slow boat. I might just solder over the
dead pixels and special case them in xy() function to disappear and live
with dead pixels for development.

I'm in the late stages of landing a commit that'll pave the way to land a
bunch more effects that would look good on a grid like that. (Some might
look better with a diffuser.) I could probably help coach you through a
build if you want to help triage what looks good and what doesn't. A remote
and/or configuring with the web server is highly recommended. Buzz me
privately (same name here as on gmail) if you would like to help with that.
That screen is 1/3 the size, much lower resolution, and a weirder aspect
ratio, but it's worth trying.

Rutger and I recently landed some changes in the doc to try to help break
the "LED_PIN0 is physical pin 5" thing in the doc that would have surely
been unhelpful for you on what I understand of your hardware. If README.md,
settings in i*/globals.h or anything else in the project can be clarified
to have reduced your pain, let's try to improve the experience for the next
person. It'll be best while it's still fresh in your mind. Once developers
are familiar enough with a project, they tend to be less good at
writing/maintaining starting guides because it's hard to spot missing
things when you already know them. (Nothing personal to our own elders
here. Just statistically true in my experience...)

It's good to know someone is actively workgin with M5Stick - in another
window just minutes ago, I considered a change that might have broken that
and I chickened out since I couldn't test it. I'll come back to that
later...or I may bring up the code on an ESP32-BOX that I'd forgotten I
have which would have also hit the ESP32-S3 + tiny LCD corner of the test
matrix . A core s3 is a bit pricey for an uncommitted test project like
that.


Great progress! I look forward to hearing more about your journey with this
code and your future projects.


RJL

**Comment by prschguy1** [comment]
Thank you for your words Robert. The display I had designed and 3d printed. It uses  standard 16x16 by 3 panels. It consists of a housing, bezel, grid, aluminum heat sink and a 3 layer diffuser. Originally I had printed it all in PLA, however it was running a bit too bright one day, and I had a melt in the grid. I have since broken the grid into 3 pieces so it could be printed using a more heat resistant resin using a sla printer. Grid was designed to be epoxied into one part once cured. As my display and chip set are still in flux, have not finished a base for it yet. Still more to come on this project and Atomlight. Definitely a good project to keep me engaged and learning on. 

Craig

**Comment by prschguy1** [comment]
![20230706_034425_resized 254](https://github.com/PlummersSoftwareLLC/NightDriverStrip/assets/69419979/9096ca0f-23c0-44a0-85f3-1dcd589c5096)
![20230707_032751_resized 255](https://github.com/PlummersSoftwareLLC/NightDriverStrip/assets/69419979/28853556-5af2-4924-b55d-2478b8750bce)
![20230707_034135_resized 257](https://github.com/PlummersSoftwareLLC/NightDriverStrip/assets/69419979/bde9bfe9-b1bc-42f4-87b0-604f3c76f45e)

**Comment by robertlipe** [comment]
Impressive! That's well worthy of an entry on the 'show and tell' part of
the Discussions group on its own. There's probably a Hackaday/Instructables
(one tends to link to the other these days) entry there, too.

Do you find the passive aluminum cooling blocks helpful? Do they help get
the heat away from the panels where just getting the heat away from the
individual bulbs would be enough? I'm not a thermals nerd, but it seems
that by the time the heat could travel the 6-8 inches to get from the
corners/edge to those center block, you'd have already lost the battle. ("I
had some laying around and just gave it a shot." is a perfectly reasonable
response - and one I'd probably issue if I had them laying around myself
and had already got the aluminum panels cut and fitted.)

I'm with you on the default brightness.  The default being the intensity of
the sun is surely appropriate for an outdoor fixture to be seen from afar,
but on a workbench, it's easy to lose an eyeball and melt something just
from the default brightness.

My own local work trees tend to have lots of hacky things like:
             set_max_power_in_milliwatts(POWER_LIMIT_MW);                //
Set
brightness limit
+//            set_max_power_in_milliwatts(1000);                // Set
brightne
ss limit
         #endif

         g_Values.Brightness = 255;
+//        g_Values.Brightness = 16;
+//        g_Values.Brightness = 1;
and
+    #define POWER_LIMIT_MW       2 * 1000   // DO NOT COMMIT

style hacks just to help prevent burning carpets, random papers, and my
retinas. :-)

I also tend to just yank the knob on the power supply and undervolt strips
to find the point where the cells (that aren't red) will just barely light
and run them like that. When I'm running 300 lights at my desk, running
them barely visible is fine.

...until I get the mysteries of the "colorserver" port solved and can just
curl a BMP of the current image and shove it into a webview whatever
appropriate way there is to render pixels of arbitrary geometry these days.
I've started versions in Qt and PHP (no judgement!) and just never really
got anywhere with a realtime viewer. Another project that needs some
attention from the incoming class! :-)

Nice work!

**Comment by prschguy1** [comment]
Hello Robert, I really think it does help having the plate and heat sinks. It is thermally coupled to the panels with thermal tape. My thoughts were that not all pixels burn bright at the same time, and in this style the heat would be more evenly spread throughout the panels.  While it is currently passive cooling, it is set up with a cooling duct. When I get the base completed, my intention is to incorporate the power supply, and a small cooling fan there which will force air flow into the display and out of the vents. While the grid melt should no longer be an issue, was hoping that cooler pixels might have a longer life span. The diffuser certainly helps in protecting the retinas from the light levels.
![20230829_051907_resized](https://github.com/PlummersSoftwareLLC/NightDriverStrip/assets/69419979/d198662b-16ba-4007-8366-1c75e9c55cad)

**Comment by robertlipe** [comment]
So cool!
